### PR TITLE
[TokenClassification] Label realignment for subword aggregation

### LINKED ISSUE #10263: NER label re-alignment always expects B labelled first sub-words
## Environment info

- `transformers` version: 4.3.1
- Platform: Darwin-19.6.0-x86_64-i386-64bit
- Python version: 3.7.7
- PyTorch version (GPU?): 1.7.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help

- bert, tokenizers, pipelines: @LysandreJik
- trainer, maintained examples: @sgugger

## Information

Model I am using (Bert, XLNet ...): [DistilBERT fine-tuned for conll03](https://huggingface.co/elastic/distilbert-base-cased-finetuned-conll03-english)

The problem arises when using:
* [x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

1. Fine-tune a BERT model for NER/conll03 using the `run_ner.py` example script, all default values
2. Correct the label alignments, see [config.json](https://huggingface.co/elastic/distilbert-base-cased-finetuned-conll03-english/blob/main/config.json)
3. Infer using entities that have not been seen at training time, and are composed of multiple word-parts as defined by WordPiece (my assumption as to the cause).
4. Sub-words are labelled but pipeline re-grouping/label alignment relies on perfect sub-word labelling:

E.g. Accenture → A ##cc ##ent ##ure → B-ORG O O O → A (ORG)
E.g. Max Mustermann → Max Must ##erman ##n → B-PER I-PER I-PER O → Max Musterman (PER)
E.g. Elasticsearch → El ##astic ##sea #rch → O O I-MISC O → ##sea (MISC)

## Expected behavior

I would expect that the realignment takes the label from the first word part or the best scoring sub-word part and propogates that label to the entire word, never returning sub-words. The default in `run_ner.py` is to use a padded sub-word label at training as per the BERT paper, but I've not tried setting that to `False` yet as that's not the typical/standard practice.

E.g. Accenture → A ##cc ##ent ##ure → B-ORG O O O → Accenture (ORG)
E.g. Max Mustermann → Max Must ##erman ##n → B-PER I-PER I-PER O → Max Mustermann (PER)
E.g. Elasticsearch → El ##astic ##sea #rch → O O I-MISC O → Elasticsearch (MISC)

I'll add that it seems odd that this business logic is in the `pipeline`. When evaluating on conll03, I assume we are using the sub-words/first word, but this realignment should be considered during evaluation. As-is, I suspect the recall is lower than it should be.

### PR DESCRIPTION
**Pull request by francescorubbo**
# What does this PR do?

Fixes #10263, #10763
See also #10568

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@LysandreJik
@sgugger 

## What capabilities have been added ?
label realignment: token predictions for subwords can be realigned with 4 different strategies

`first` (default): the prediction for the first token in the word is assigned to all subword tokens
`max`: the highest confidence prediction among the subword tokens is assigned to all subword tokens
`average`: the average pool of the predictions for all subwords is assigned to all subword tokens

## What are the expected changes from the current behavior?
New flag `aggregation_strategy` enables realignment.
Already existing flag `ignore_subwords` actually enables merging subwords.

## Example use cases with code sample enabled by the PR
```
ner = transformers.pipeline(
    'ner',
    model='elastic/distilbert-base-cased-finetuned-conll03-english',
    tokenizer='elastic/distilbert-base-cased-finetuned-conll03-english',
    ignore_labels=[],
    ignore_subwords=False,
    aggregation_strategy='average'
)
ner('Mark Musterman')
[
    {
        'word': 'Mark',
        'score': 0.999686598777771,
        'index': 1,
        'start': 0,
        'end': 4,
        'is_subword': False,
        'entity': 'B-PER'
    },
    {
        'word': 'Must',
        'score': 0.9995412826538086,
        'index': 2,
        'start': 5,
        'end': 9,
        'is_subword': False,
        'entity': 'I-PER'
    },
    {
        'word': '##erman',
        'score': 0.9996127486228943,
        'index': 3,
        'start': 9,
        'end': 14,
        'is_subword': True,
        'entity': 'I-PER'
    }
]
ner = transformers.pipeline(
    'ner',
    model='elastic/distilbert-base-cased-finetuned-conll03-english',
    tokenizer='elastic/distilbert-base-cased-finetuned-conll03-english',
    ignore_labels=[],
    ignore_subwords=True,
    aggregation_strategy='average'
)
ner('Mark Musterman')
[
    {
        'word': 'Mark',
        'score': 0.999686598777771,
        'index': 1,
        'start': 0,
        'end': 4,
        'is_subword': False,
        'entity': 'B-PER'
    },
    {
        'word': 'Musterman',
        'score': 0.9995412826538086,
        'index': 2,
        'start': 5,
        'end': 9,
        'is_subword': False,
        'entity': 'I-PER'
    }
]
```

## Previous use cases with code sample that see the behavior changes
```
ner = transformers.pipeline(
    'ner',
    model='elastic/distilbert-base-cased-finetuned-conll03-english',
    tokenizer='elastic/distilbert-base-cased-finetuned-conll03-english',
    ignore_labels=[],
    ignore_subwords=True
)
ner('Mark Musterman')
[
    {
        'word': 'Mark',
        'score': 0.999686598777771,
        'entity': 'B-PER',
        'index': 1,
        'start': 0,
        'end': 4
    },
    {
        'word': 'Must',
        'score': 0.9995412826538086,
        'entity': 'I-PER',
        'index': 2,
        'start': 5,
        'end': 9
    },
    {
        'word': '##erman',
        'score': 0.9996127486228943,
        'entity': 'I-PER',
        'index': 3,
        'start': 9,
        'end': 14
    }
]
```

### CODE REVIEW

File: tests/test_pipelines_token_classification.py [hunk 1]
```diff
@@ -190,6 +191,329 @@ def _test_pipeline(self, nlp: Pipeline):
             ],
         ]
 
+        ungrouped_inputs_all_scores = [
```
Comment (sgugger) [inline tests/test_pipelines_token_classification.py:194]: Should this be in a fixture too?

**Review by sgugger — APPROVED** [review]
Thanks for re-opening a clean PR!

File: tests/test_pipelines_token_classification.py (hunk 1 above)
Comment (francescorubbo) [reply tests/test_pipelines_token_classification.py]: Good point! Moved to json fixture along w/ the already existing inputs above this.

**Comment by francescorubbo** [comment]
@sgugger the failed test (`test_gpt2_model_past_large_inputs`) seems unrelated to the changes in this PR. Any thought on what might be going on and how to resolve?

**Comment by sgugger** [comment]
No, it's just flaky, don't worry!

**Review by LysandreJik — APPROVED** [review]
Thank you for your work! This LGTM.

I'm putting links to relevant comments from the previous PR:
- This keeps backwards compatibility https://github.com/huggingface/transformers/pull/10568#pullrequestreview-627913247
- Completing the work that was started here, left for a future PR https://github.com/huggingface/transformers/pull/10568#issuecomment-810052082

@Narsil could you give it a quick look if you have time? It shouldn't change anything to the current pipeline behavior, but offers a cleaner `AggregationStrategy` as an opt-in.

**Comment by LysandreJik** [comment]
Hey @francescorubbo, @Narsil pointed out a few issues with the current implementation that we'll take a look at today/tomorrow. Namely, the code becomes a bit complex as we keep adding features to that pipeline so it might be time for a slightly larger refactor, and some code is model-specific, such as this line which wouldn't work on non BERT-like tokenizers:

```
subwords[0]["word"] += "".join([sub["word"].split("##")[1] for sub in subwords[1:]])
```

We're taking a look at what can be done and will come back to you in a bit. Thanks again for your patience.
